<a href="https://colab.research.google.com/github/Arjun-R-krishnan/-Supervised-learning/blob/main/supervised_learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Importing necessary libraries

In [1]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_breast_cancer
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, f1_score, accuracy_score

## Loading the dataset

In [2]:
cancer =  load_breast_cancer()

In [3]:
cancer

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

## Converting the data to a dataframe for easy understanding

In [4]:
cancer_df = pd.DataFrame(cancer.data,columns = cancer.feature_names)

In [5]:
# printing the dataframe
cancer_df.head(5)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

## Adding the target column to the datfarame

In [6]:
cancer_df['Result'] = cancer.target

In [7]:
#checking the dataframe to ensure the new column is placed
cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  Result  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625          0.2364                  0.076

In [8]:
# checking the shape of the dataframe
cancer_df.shape

(569, 31)

In [9]:
#showing the summary of the dataframe
cancer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [10]:
#checking for null values
cancer_df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
Result                     0
dtype: int64

In [11]:
cancer_df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension      Result  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [12]:
# checking the counts of resultant values
cancer_df['Result'].value_counts()

Result
1    357
0    212
Name: count, dtype: int64

 1 --> banign     : are non cancerous tumors

 0 --> Malignant  : are cancerous tumors

## Storing the input variables to x

In [13]:
# Dropping the 'Result' column from the DataFrame 'cancer_df' and storing the remaining columns in 'x'
x = cancer_df.drop(columns = 'Result' , axis = 1)

In [14]:
x.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

## Storing the output variable to y

In [15]:
# Storing the 'Result' column from the DataFrame 'cancer_df' in the variable 'y
y = cancer_df['Result']

In [16]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Result, dtype: int64

## Partitioning the dataset into training and testing subsets

In [17]:
# 80% of the data will be used for training and 20% for testing
x_train ,x_test ,y_train ,y_test = train_test_split(x,y,test_size = 0.2,random_state= 2)

In [18]:
#checking the no of rows of the training data
x_train.shape

(455, 30)

In [19]:
# checking the no of rows of testing data
x_test.shape

(114, 30)

## Applying feature scaling using StandardScaler to normalize the dataset



In [20]:
# Initializing the StandardScaler to standardize features by removing the mean and scaling to unit variance
scaler = StandardScaler()

In [21]:
# Applying standard scaling to the training data and converting it back to a DataFrame with original column names
x_train = pd.DataFrame(scaler.fit_transform(x_train),columns=x_train.columns)

In [22]:
# Applying standard scaling to the test data and converting it back to a DataFrame with original column names
x_test = pd.DataFrame(scaler.fit_transform(x_test),columns=x_test.columns)

In [23]:
x_train.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0    -0.013303      1.775766       -0.014920  -0.144789         0.228790   
1    -0.844828     -0.628428       -0.877027  -0.765610        -1.019821   
2     1.447559      0.711802        1.474288   1.405424        -0.758582   
3    -0.081648      0.999483       -0.007498  -0.152149         2.546938   
4    -0.896086      0.451519       -0.818069  -0.793920         1.872772   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0          0.175467       -0.544402            -0.128256      -1.012054   
1         -1.249059       -0.944554            -0.889091      -1.109505   
2          0.754792        0.319555             0.511737      -0.149990   
3          1.407012        0.893973             1.135134       1.664093   
4          0.351950        0.228993             0.225711       0.329767   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0               -0.152706  ...     -0.186400       1.187213        -0.196141   
1               -0.186165  ...     -0.949992      -0.884569        -0.980765   
2               -1.086665  ...      1.513753       1.008666         1.813225   
3                1.665752  ...      0.636255       2.583414         0.662848   
4                0.864176  ...     -0.705305       1.536263        -0.590122   

   worst area  worst smoothness  worst compactness  worst concavity  \
0   -0.291354         -0.332182          -0.171367        -0.659330   
1   -0.806497         -1.240630          -1.224541        -1.078200   
2    1.478636          0.016558           1.968844         0.565914   
3    0.513646          3.908676           1.717432         0.942800   
4   -0.638931          2.543851           0.106982         0.642160   

   worst concave points  worst symmetry  worst fractal dimension  
0             -0.132370       -1.080145                -0.035279  
1             -1.115526       -0.857740                -0.720989  
2              0.875840        0.496760                 0.463217  
3              1.341755        2.568643                 1.371035  
4              0.442004        0.113821                 0.676463  

[5 rows x 30 columns]

In [24]:
x_test.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0    -0.087003     -1.471929       -0.105374  -0.214797         2.056279   
1     0.199891      0.035773        0.170618   0.055316        -0.587426   
2    -1.288584     -0.218477       -1.306678  -1.108112        -1.103585   
3     2.255967      0.180696        2.461104   2.439328         2.647475   
4     0.048006     -0.615107        0.052336  -0.078435        -0.012147   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0         -0.187598        0.043460             0.343147       0.486932   
1         -0.175255       -0.429668            -0.113156       0.163461   
2         -0.973011       -0.937028            -1.190931      -0.500154   
3          3.066046        3.744315             3.294129       2.421088   
4          0.109178       -0.194582            -0.264580       0.273508   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                0.209715  ...     -0.386349      -1.696507        -0.421900   
1               -0.572971  ...     -0.069972       0.165328        -0.015412   
2                0.151499  ...     -1.059863      -0.105077        -1.074567   
3                0.962646  ...      2.288300       0.478792         2.708447   
4                0.151499  ...      0.011549      -0.640721         0.048711   

   worst area  worst smoothness  worst compactness  worst concavity  \
0   -0.445575          0.230418          -0.755219        -0.601924   
1   -0.193459         -0.156517           0.272404        -0.486802   
2   -0.885530          0.201756          -0.822452        -1.033324   
3    2.303861          1.696952           2.719329         3.180175   
4   -0.109205         -0.247280           0.482739         0.429942   

   worst concave points  worst symmetry  worst fractal dimension  
0             -0.266292       -1.097764                -0.655975  
1              0.448441        0.060666                 0.021082  
2             -1.419815        0.197886                -0.310504  
3              2.591129        1.639423                 0.752569  
4              0.335350        0.192109                 0.183264  

[5 rows x 30 columns]

In [25]:
x_test.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0    -0.087003     -1.471929       -0.105374  -0.214797         2.056279   
1     0.199891      0.035773        0.170618   0.055316        -0.587426   
2    -1.288584     -0.218477       -1.306678  -1.108112        -1.103585   
3     2.255967      0.180696        2.461104   2.439328         2.647475   
4     0.048006     -0.615107        0.052336  -0.078435        -0.012147   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0         -0.187598        0.043460             0.343147       0.486932   
1         -0.175255       -0.429668            -0.113156       0.163461   
2         -0.973011       -0.937028            -1.190931      -0.500154   
3          3.066046        3.744315             3.294129       2.421088   
4          0.109178       -0.194582            -0.264580       0.273508   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                0.209715  ...     -0.386349      -1.696507        -0.421900   
1               -0.572971  ...     -0.069972       0.165328        -0.015412   
2                0.151499  ...     -1.059863      -0.105077        -1.074567   
3                0.962646  ...      2.288300       0.478792         2.708447   
4                0.151499  ...      0.011549      -0.640721         0.048711   

   worst area  worst smoothness  worst compactness  worst concavity  \
0   -0.445575          0.230418          -0.755219        -0.601924   
1   -0.193459         -0.156517           0.272404        -0.486802   
2   -0.885530          0.201756          -0.822452        -1.033324   
3    2.303861          1.696952           2.719329         3.180175   
4   -0.109205         -0.247280           0.482739         0.429942   

   worst concave points  worst symmetry  worst fractal dimension  
0             -0.266292       -1.097764                -0.655975  
1              0.448441        0.060666                 0.021082  
2             -1.419815        0.197886                -0.310504  
3              2.591129        1.639423                 0.752569  
4              0.335350        0.192109                 0.183264  

[5 rows x 30 columns]

## Algorithm Implimentation

### Logistic Regression

How it works


Logistic Regression models the probability that an instance belongs to a particular class (binary classification in this case: benign vs malignant) using a sigmoid function to convert the linear output into a probability. It assumes a linear relationship between the independent variables and the log-odds of the dependent variable.


Why suitable



It is interpretable and efficient for binary classification problems like this one. Logistic Regression works well when the features have a linear relationship with the output, making it a good starting point.

In [26]:
# Creating an instance of the LogisticRegression class
lr_model = LogisticRegression()

In [27]:
# Training the Logistic Regression model using the training data
lr_model.fit(x_train,y_train)

LogisticRegression()

In [28]:
# Making predictions on the test dataset
y_pred_lr = lr_model.predict(x_test)

In [29]:
# Generate classification report
lr_report = classification_report(y_test, y_pred_lr, output_dict=True)

In [30]:
lr_report_df = pd.DataFrame(lr_report).transpose()

In [31]:
lr_report_df

precision    recall  f1-score     support
0              1.000000  0.933333  0.965517   45.000000
1              0.958333  1.000000  0.978723   69.000000
accuracy       0.973684  0.973684  0.973684    0.973684
macro avg      0.979167  0.966667  0.972120  114.000000
weighted avg   0.974781  0.973684  0.973510  114.000000

In [32]:
# Calculating the accuracy of the Logistic Regression model
accuracy_of_lr_model = accuracy_score(y_test,y_pred_lr)

In [33]:
print(f"The accuracy of the logistic regression model is {accuracy_of_lr_model:.2f}")

The accuracy of the logistic regression model is 0.97


##  Decision Tree Classifier


How it works

 This algorithm splits the data into subsets based on the value of features using conditions that maximize information gain or minimize impurity (Gini or entropy). It builds a tree where each internal node represents a decision on a feature, and each leaf represents a class label.

Why suitable

Decision Trees can handle both numerical and categorical data and are easy to visualize. They are non-linear models, so they can capture complex interactions between features. Since the breast cancer dataset has many features, this model can identify important ones and perform well.


In [34]:
# Creating an instance of the Decisiom tree classifier class
dtc_model = DecisionTreeClassifier()

In [35]:
# Training the Deciion tree classifier model using the training data
dtc_model.fit(x_train,y_train)

DecisionTreeClassifier()

In [36]:
# Making predictions on the test dataset
y_pred_dtc = dtc_model.predict(x_test)

In [37]:
# Generate classification report
dtc_report = classification_report(y_test, y_pred_dtc, output_dict=True)

In [38]:
dtc_report_df = pd.DataFrame(dtc_report).transpose()

In [39]:
dtc_report_df

precision    recall  f1-score     support
0              0.895833  0.955556  0.924731   45.000000
1              0.969697  0.927536  0.948148   69.000000
accuracy       0.938596  0.938596  0.938596    0.938596
macro avg      0.932765  0.941546  0.936440  114.000000
weighted avg   0.940540  0.938596  0.938905  114.000000

In [40]:
# Calculating the accuracy of the Decision tree classifier  model
accuracy_of_dtc_model = accuracy_score(y_test,y_pred_dtc)

In [41]:
print(f"The accuracy of the Decision Tree Classifier  model is {accuracy_of_dtc_model:.2f}")

The accuracy of the Decision Tree Classifier  model is 0.94


##  Random Forest Classifier


How it works

Random Forest is an ensemble learning method that builds multiple decision trees and combines their predictions (via averaging for regression or majority vote for classification). Each tree is built using a random subset of features, reducing overfitting

Why suitable

Random Forest is robust to overfitting and handles high-dimensional datasets well, making it suitable for the breast cancer dataset with 30 features. It also gives insight into feature importance.

In [42]:
# Creating an instance of the Random Forest Classifier class
rfc_model = RandomForestClassifier()

In [43]:
# Training the Random Forest  classifier model using the training data
rfc_model.fit(x_train,y_train)

RandomForestClassifier()

In [44]:
# Making predictions on the test dataset
y_pred_rfc = rfc_model.predict(x_test)

In [45]:
# Generate classification report
rfc_report = classification_report(y_test, y_pred_rfc, output_dict=True)

In [46]:
rfc_report_df = pd.DataFrame(rfc_report).transpose()

In [47]:
rfc_report_df

precision    recall  f1-score     support
0              0.931818  0.911111  0.921348   45.000000
1              0.942857  0.956522  0.949640   69.000000
accuracy       0.938596  0.938596  0.938596    0.938596
macro avg      0.937338  0.933816  0.935494  114.000000
weighted avg   0.938500  0.938596  0.938472  114.000000

In [48]:
# Calculating the accuracy of the Random forest  classifier  model
accuracy_of_rfc_model = accuracy_score(y_test,y_pred_rfc)

In [49]:
print(f"The accuracy of the Random Forest  Classifier  model is {accuracy_of_rfc_model:.2f}")

The accuracy of the Random Forest  Classifier  model is 0.94


## Support Vector Machine


How it works

SVM tries to find a hyperplane that best separates the data points into two classes with the largest margin. It uses kernel tricks to handle non-linear relationships between features and the output.

Why suitable

SVM is effective in high-dimensional spaces and works well when the classes are separable.

In [50]:
# Creating an instance of the Support Vector Machine class
svc_model = SVC()

In [51]:
# Training the SVM model using the training data
svc_model.fit(x_train,y_train)

SVC()

In [52]:
# Making predictions on the test dataset
y_pred_svc= svc_model.predict(x_test)

In [53]:
# Generate classification report
svc_report = classification_report(y_test, y_pred_svc, output_dict=True)

In [54]:
svc_report_df = pd.DataFrame(svc_report).transpose()

In [55]:
svc_report_df

precision    recall  f1-score     support
0              0.977273  0.955556  0.966292   45.000000
1              0.971429  0.985507  0.978417   69.000000
accuracy       0.973684  0.973684  0.973684    0.973684
macro avg      0.974351  0.970531  0.972355  114.000000
weighted avg   0.973735  0.973684  0.973631  114.000000

In [56]:
# Calculating the accuracy of the SVM  model
accuracy_of_svm_model = accuracy_score(y_test,y_pred_svc)

In [57]:
print(f"The accuracy of the SVM  model is {accuracy_of_svm_model:.2f}")

The accuracy of the SVM  model is 0.97


##  k-Nearest Neighbors (k-NN)


How it works

 k-NN classifies a data point based on the majority class of its 'k' nearest neighbors in the feature space. Distance measures like Euclidean distance are used to find neighbors

Why suitable

k-NN is simple and effective, especially for small to medium-sized datasets. While it may be slower with large datasets, it can still work well here because the dataset isn't too large (569 instances), and the features are numeric, which is ideal for distance-based classification.

In [58]:
# Creating an instance of the KNN class
knn_model =  KNeighborsClassifier()

In [59]:
# Training the KNN model using the training data
knn_model.fit(x_train,y_train)

KNeighborsClassifier()

In [60]:
# Making predictions on the test dataset
y_pred_knn= knn_model.predict(x_test)

In [61]:
# Generate classification report
knn_report = classification_report(y_test, y_pred_knn, output_dict=True)

In [62]:
knn_report_df = pd.DataFrame(knn_report).transpose()

In [63]:
knn_report_df

precision    recall  f1-score     support
0              1.000000  0.933333  0.965517   45.000000
1              0.958333  1.000000  0.978723   69.000000
accuracy       0.973684  0.973684  0.973684    0.973684
macro avg      0.979167  0.966667  0.972120  114.000000
weighted avg   0.974781  0.973684  0.973510  114.000000

In [64]:
# Calculating the accuracy of the KNN model
accuracy_of_knn_model = accuracy_score(y_test,y_pred_knn)

In [65]:
print(f"The accuracy of the KNN model is {accuracy_of_knn_model:.2f}")

The accuracy of the KNN model is 0.97


#### Since the accuracy of the three algorithms are same we are using class 1 F1 score to compare the perfomance of the ml algorithms

In [66]:
# Create a list of classification report dataframes for comparison across models
report_dfs = [lr_report_df,dtc_report_df,rfc_report_df,svc_report_df,knn_report_df]

In [67]:
# Create a list of model names corresponding to the classification reports
model_names = ['Logistic Regression','Decision Tree','Random Forest','SVM','KNN']

In [68]:
# Initialize an empty list to store the F1-scores for class 1 across different models
f1_scores_class_1 = []

In [69]:
# Loop through each dataframe in the report_dfs list to extract the F1-scores for class 1
for df in report_dfs:
    f1_score = df.iloc[1, 2]
    f1_scores_class_1.append(f1_score)

In [70]:
f1_scores_class_1

[0.9787234042553191,
 0.9481481481481482,
 0.9496402877697842,
 0.9784172661870504,
 0.9787234042553191]

In [71]:
# Create a dictionary that maps model names to their corresponding F1-scores for class 1
model_f1_scores = {model_names[i]: f1_scores_class_1[i] for i in range(len(model_names))}

In [72]:
model_f1_scores

{'Logistic Regression': 0.9787234042553191,
 'Decision Tree': 0.9481481481481482,
 'Random Forest': 0.9496402877697842,
 'SVM': 0.9784172661870504,
 'KNN': 0.9787234042553191}

In [73]:
df_f1_scores = pd.DataFrame(list(model_f1_scores.items()), columns=['Model', 'F1 Score'])

In [74]:
df_f1_scores

Model  F1 Score
0  Logistic Regression  0.978723
1        Decision Tree  0.948148
2        Random Forest  0.949640
3                  SVM  0.978417
4                  KNN  0.978723

## Finding the model with high perfomance

In [75]:
# Find the key corresponding to the best f1 score
max_key = max(model_f1_scores, key=model_f1_scores.get)
print(f"Model with the highest F1-Score for Class 1: {max_key}")


Model with the highest F1-Score for Class 1: Logistic Regression


In [76]:
## Finding the model with low perfomance

In [77]:
# Find the key corresponding to the worst f1 score
min_key = min(model_f1_scores, key=model_f1_scores.get)
print(f"Model with the lowest F1-Score for Class 1: {min_key}")


Model with the lowest F1-Score for Class 1: Decision Tree
